File: backend/src/main/java/org/luxons/sevenwonders/repositories/LobbyRepository.java
package org.luxons.sevenwonders.repositories;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

import org.luxons.sevenwonders.game.data.GameDefinitionLoader;
import org.luxons.sevenwonders.lobby.Lobby;
import org.luxons.sevenwonders.lobby.Player;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Repository;

@Repository
public class LobbyRepository {

    private final GameDefinitionLoader gameDefinitionLoader;

    private Map<Long, Lobby> lobbies = new HashMap<>();

    private long lastGameId = 0;

    @Autowired
    public LobbyRepository(GameDefinitionLoader gameDefinitionLoader) {
        this.gameDefinitionLoader = gameDefinitionLoader;
    }

    public Collection<Lobby> list() {
        return lobbies.values();
    }

    public Lobby create(String gameName, Player owner) {
        long id = lastGameId++;
        Lobby lobby = new Lobby(id, gameName, owner, gameDefinitionLoader.getGameDefinition());
        lobbies.put(id, lobby);
        return lobby;
    }

    public Lobby find(long lobbyId) throws LobbyNotFoundException {
        Lobby lobby = lobbies.get(lobbyId);
        if (lobby == null) {
            throw new LobbyNotFoundException(lobbyId);
        }
        return lobby;
    }

    public Lobby remove(long lobbyId) throws LobbyNotFoundException {
        Lobby lobby = lobbies.remove(lobbyId);
        if (lobby == null) {
            throw new LobbyNotFoundException(lobbyId);
        }
        return lobby;
    }

    public static class LobbyNotFoundException extends RuntimeException {
        LobbyNotFoundException(long id) {
            super("Lobby not found for id '" + id + "'");
        }
    }
}

File: backend/src/main/kotlin/org/luxons/sevenwonders/repositories/LobbyRepository.kt
package org.luxons.sevenwonders.repositories

import org.luxons.sevenwonders.game.data.GameDefinitionLoader
import org.luxons.sevenwonders.lobby.Lobby
import org.luxons.sevenwonders.lobby.Player
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Repository
import java.util.HashMap

@Repository
class LobbyRepository @Autowired constructor() {

    private val gameDefinitionLoader: GameDefinitionLoader = GameDefinitionLoader()

    private val lobbies = HashMap<Long, Lobby>()

    private var lastGameId: Long = 0

    fun list(): Collection<Lobby> = lobbies.values

    fun create(gameName: String, owner: Player): Lobby {
        val id = lastGameId++
        val lobby = Lobby(id, gameName, owner, gameDefinitionLoader.gameDefinition)
        lobbies[id] = lobby
        return lobby
    }

    fun find(lobbyId: Long): Lobby = lobbies[lobbyId] ?: throw LobbyNotFoundException(lobbyId)

    fun remove(lobbyId: Long): Lobby = lobbies.remove(lobbyId) ?: throw LobbyNotFoundException(lobbyId)
}

internal class LobbyNotFoundException(id: Long) : RuntimeException("Lobby not found for id '$id'")
